src/CodeGeneration.Roslyn/DocumentTransform.cs

```diff
@@ -51,16 +50,19 @@ public static class DocumentTransform
 
             var inputFileLevelUsingDirectives = inputSyntaxTree.GetRoot().ChildNodes().OfType<UsingDirectiveSyntax>();
 
-            var memberNodes = from syntax in inputSyntaxTree.GetRoot().DescendantNodes(n => n is CompilationUnitSyntax || n is NamespaceDeclarationSyntax || n is TypeDeclarationSyntax).OfType<MemberDeclarationSyntax>()
-                              select syntax;
+            var memberNodes = inputSyntaxTree.GetRoot().DescendantNodesAndSelf().OfType<CSharpSyntaxNode>();
```

Comment:
Isn't removing descdentant filter going to cause traversal of every single syntax node below Type level? It's probably going to hurt performance badly, taking into account laziness of Roslyn's tree model; and I can't see why it'd ever be useful/necessary.

Reply:
@amis92 this change was nessessary e.g. becaue `OfType<MemberDeclarationSyntax>` wasn't returning `CompilationUnitSyntax` or `NamespaceDeclarationSyntax` because they don't derive one from another. Another answer - user may want to attribute something else (e.g. property or field) and get generation based on it. Why this scenario should be not supported?

Code generation doesn't have often in my practice, there is often a separated assembly for all generated stuff, so it shouldn't regenerate a lot, however, it provides more flexibility. 

So my question: library should provide the widest support unless pefromance degradation isn't notable

Reply:
Of course, I didn't contest `OfType` change, that I understand. Rather the `Descendants` change: right now we visit absolutely everything, including method bodies.

I think Properties and Fields might be good targets for Attributes, but on the other hand I believe a class that generates code should be decorated itself. And if it's a class that triggers generation, maybe better to leave the rest to the generator?

This is a big change from current behavior, and I think it's worth a separate discussion+PR.

Reply:
Ok, it may be pushed as separated PR if/when we have need in it.